### PR TITLE
Update Avro schema parsing to be identical everywhere

### PR DESCRIPTION
**Pull request by quodlibetor**
This supports using Schema Registry URLs, Key Schemas and inline/file schemas in the two places that
they can be used. This does *not* change the syntax of `ENVELOPE UPSERT`, which still allows
specifying an extra `FORMAT AVRO` after itself, a later PR will deprecate that and use the new
unified syntax.

This allows optionally specifying Avro inline value schemas as `VALUE SCHEMA` instead of just
`SCHEMA` or requiring `VALUE SCHEMA`, to merge the difference between the CSR syntax and the
inline-only syntax.

This should correctly populate the key schema from the CSR if the schema exists, I'll test that in
the follow-up PR where we start actually using this data.

This doesn't change any semantics of any other part of the system, and detailed testing will come
with the follow-up PR, things are left as-is for now to ensure that this doesn't change any
behavior.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/6286)

### CODE REVIEW

File: src/sql-parser/src/parser.rs [hunk 1]
```diff
@@ -1564,19 +1554,35 @@ impl<'a> Parser<'a> {
...
     }
 
+    fn parse_schema_pair(&mut self) -> Result<SchemaPair, ParserError> {
+        let key_schema = if self.parse_keyword(KEY) {
+            Some(self.parse_schema()?)
+        } else {
+            None
+        };
+        if !self.parse_keyword(VALUE) {
+            warn!("Avro schemas will require the syntax 'VALUE SCHEMA' in a future release");
```
Comment (benesch) [inline src/sql-parser/src/parser.rs:1577]: The chance that this warning is seen by the person who actually has the capacity to fix the query is somewhat slim. I guess this is better than nothing but it's too bad we can't easily send back a notice response to the client. 

I think we could at least make this message a bit more specific; maybe something like "the `FORMAT AVRO SCHEMA ...` syntax in `CREATE SOURCE` and `CREATE SINK` is deprecated; a future release will require that you explicitly indicate the schema types, as in `FORMAT AVRO [KEY SCHEMA ...] VALUE SCHEMA ...`.

File: src/sql/src/plan/statement/ddl.rs [hunk 2]
```diff
@@ -1027,6 +1023,22 @@ pub fn plan_create_source(
     })
 }
 
+/// Take the inline schemas that have been purified out
+///
+/// Panics if purify doesn't correclty inline schemas
```
Comment (benesch) [inline src/sql/src/plan/statement/ddl.rs:1028]: I think this comment could use another pass. "Take the inline schemas that have been purified out" is IMO both more verbose and less clear than the name of the function ("extract inline schemas"). Also "s/correclty/correctly/" .

**Review by benesch — CHANGES_REQUESTED** [review]
I think I'm missing the context on the broader goal here. Are we just trying to make it possible to specify key and value schemas explicitly? Or is this in service of some larger effort?

I did a detailed review and left comments within, but maybe hold off on those for a moment because I ended up with some meta questions.

This seems like it doesn't play well with upsert sources? This is now syntactically permitted, but seems really weird:

```
CREATE SOURCE src
FROM KAFKA BROKER 'k' TOPIC 't'
FORMAT AVRO USING KEY SCHEMA ... VALUE SCHEMA ...
ENVELOPE UPSERT FORMAT AVRO KEY SCHEMA ... VALUE SCHEMA ...
```

I think maybe the offender here is `UPSERT`? I think upsert's Avro format needs to be somehow limited to `FORMAT AVRO SCHEMA ...` (i.e., a single schema). It doesn't really make sense for upsert schemas to talk about both a key and value schema, since upsert only cares about the key schema.

Part of the tension here is that Avro schemas are so special. Like, it doesn't actually make sense that `FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY` surfaces up the key schema, because that format spec should really just be about the value schema. What we really wanted was probably something like:

```
VALUE FORMAT {AVRO|...}
KEY FORMAT {AVRO|...}
[ENVELOPE UPSERT]
```

The current design is kind of a hack to extract primary key information from Debezium style sources.

I'm starting to think we ought to do a design doc for this. I'm worried about further confusing the whole `FORMAT`/`ENVELOPE` thing in `CREATE SOURCE`. I think there's probably a nice path forward here that actually improves the clarity of `CREATE SOURCE` _and_ solves your problem—but I'd definitely need the context that'd be provided by a design doc to help.

File: src/sql-parser/src/ast/defs/ddl.rs [hunk 3]
```diff
@@ -56,11 +56,11 @@ impl_display!(Schema);
 pub enum AvroSchema<T: AstInfo> {
     CsrUrl {
         url: String,
-        seed: Option<CsrSeed>,
+        seed: Option<SchemaPair>,
```
[the remark this answers is not in the document]
Comment (quodlibetor) [reply src/sql-parser/src/ast/defs/ddl.rs:59]: Why is it critical here and not the other places that we accept files for schema definitions? This seems like something that should have been documented with a code comment.

File: src/sql-parser/src/parser.rs (hunk 1 above)
Comment (quodlibetor) [reply src/sql-parser/src/parser.rs:1577]: I agree a notice would be better, do we have any infrastructure for that?

**Comment by quodlibetor** [comment]
> I'm starting to think we ought to do a design doc for this. I'm worried about further confusing the whole FORMAT/ENVELOPE thing in CREATE SOURCE.

Right, I will write up a design doc. I actually addressed several of these issues in the commit message, but forgot to update the PR description to provide the context. Description is updated now, but a design doc is forthcoming.

**Comment by quodlibetor** [comment]
converting to draft until we agree on the design doc.

File: src/sql-parser/src/parser.rs (hunk 1 above)
Comment (quodlibetor) [reply src/sql-parser/src/parser.rs:1577]: This is also an attempt to combat the problem of warn-level visibility: https://github.com/MaterializeInc/materialize/pull/6311

**Comment by quodlibetor** [comment]
Implemented in #6487

**Pull request closed without merging**